R/pmcmc_parameters_nested.R

```diff
@@ -0,0 +1,594 @@
+
+    if (is.na(dim(proposal_varied)[3])) {
+      proposal_varied <- array(proposal_varied,
+                               dim = c(dim(proposal_varied), len_pop))
+    }
+
+    if (!identical(dim(proposal_varied), c(len_varied, len_varied, len_pop))) {
+      stop(sprintf(
+        "Expected proposal array with dimensions %d x %d x %d.",
+        len_varied, len_varied, len_pop))
+    }
+
+    if (!is.null(dim_names_var[[3]])) {
+      if (!identical(dim_names_var[[3]], populations)) {
+        stop("Expected 3rd dimension names of 'proposal' to match
+        populations.")
+      }
+    }
+    dimnames(proposal_varied)[3] <- list(populations)
+
+    if (!is.null(dim_names_var[[1]]) || !is.null(dim_names_var[[2]])) {
+      ok <- identical(dim_names_var[[1]], params$varied) &&
+        identical(dim_names_var[[2]], params$varied)
+      if (!ok) {
+        stop("Expected row and column names of 'proposal_varied' to match
+        varied parameters.")
+      }
+    }
+    dimnames(proposal_varied)[1:2] <- list(params$varied, params$varied)
+  }
+
+  # minimal checks for fixed proposals as these will be passed to constructors
+  if (!is.null(proposal_fixed)) {
+    dim_names_fix <- dimnames(proposal_fixed)
+    assert_is(proposal_fixed, "matrix")
+    if (!is.null(dim_names_fix[[1]]) || !is.null(dim_names_fix[[2]])) {
+      ok <- identical(dim_names_fix[[1]], params$fixed) &&
+        identical(dim_names_fix[[2]], params$fixed)
+      if (!ok) {
+        stop("Expected row and column names of 'proposal_fixed' to match
+        fixed parameters.")
+      }
+    }
+    dimnames(proposal_fixed)[1:2] <- list(params$fixed, params$fixed)
+  }
+
+  list(fixed = proposal_fixed, varied = proposal_varied)
+}
+
+make_proposals <- function(kernel_varied, kernel_fixed, params) {
+  if (!is.null(kernel_varied)) {
+    kernel_varied_list <- lapply(seq_len(dim(kernel_varied)[[3]]), function(i)
+      expand(kernel_varied[, , i, drop = FALSE], params$original))
+    names(kernel_varied_list) <- dimnames(kernel_varied)[[3]]
+    proposal_varied <- lapply(kernel_varied_list, rmvnorm_generator)
+  } else {
+    proposal_varied <- NULL
+    kernel_varied_list <- NULL
+  }
+
+  if (!is.null(kernel_fixed)) {
+    proposal_fixed <- make_proposal_fixed(kernel_fixed)
+  } else {
+    proposal_fixed <- NULL
+  }
+
+  list(kernel_varied = kernel_varied_list,
+       kernel_fixed = kernel_fixed,
+       proposal_varied = proposal_varied,
+       proposal_fixed = proposal_fixed)
+
+}
+
+expand <- function(m, complete) {
+  n <- length(complete)
+  ret <- matrix(0, ncol = n, nrow = n, dimnames = list(complete, complete))
+  ret[rownames(m), colnames(m)] <- m
+  ret
+}
+
+make_proposal_fixed <- function(kernel) {
+  rmv <- rmvnorm_generator(kernel)
+  nms <- rownames(kernel)
+  function(mean, scale = 1.0) {
+    mean[, nms] <- rep(rmv(mean[1, nms], scale = scale), each = nrow(mean))
+    mean
+  }
+}
+
+get_inner_params <- function(param, population, private) {
```

Comment:
Hmm, I wonder if we might tidy this up later - some way of propagating the information back via the summary method? Doesn't need to be done now